sycl/source/detail/pi_opencl.cpp

```diff
@@ -59,9 +59,35 @@ pi_result OCL(piDevicesGet)(pi_platform      platform,
   return cast<pi_result>(result);
 }
 
+// Entry type, matches OpenMP for compatibility, used in binary selection
+struct __tgt_offload_entry {
+  void *addr;
+  char *name;
+  size_t size;
+  int32_t flags;
+  int32_t reserved;
+};
+
+static bool imageContainsKernel(pi_device_binary image,
+                                const char *kernel_name) {
+  const auto *EntriesB =
+      static_cast<__tgt_offload_entry *>(image->EntriesBegin);
+  const auto *EntriesE = static_cast<__tgt_offload_entry *>(image->EntriesEnd);
+  // If the entry table is empty, assume that the image contains the kernel
+  if (EntriesB == EntriesE)
+    return true;
+
+  for (const auto *EntriesIt = EntriesB; EntriesIt != EntriesE; ++EntriesIt) {
```

Comment:
Initially I was going to suggest using `EntriesB` for iteration, but after a bit of thought, I like the readability brought by your approach.